### ISSUE TITLE
[Ideas] Major Changes To Project

### ISSUE THREAD

**Issue by 10carfernandez** [evolvio/evolv.io#101]
Hi everyone. I wanted to apologize for not having been on here for long. I think someone had said something to me and I never replied. I've been busy because I finally got a job and have less time. 

I've made a ton of changes locally that I'm about to push to one of my branches (Vision). I know it's terrible practice to do so many changes as once, but I wanted to make it be known because I think some of these changes might be desired. 

Since I'm pretty sure that merging all of these changes will be unacceptable, I'll try to separate them to the best of my ability. Then, people can pick and choose which ones they think I should do a PR for, and I'll manually merge them individually into my main branch before doing a PR:

1. **Improved Performance**: 
Using the sigmoid function bogs down the program because you're doing a fairly hefty computation for just a relatively small range of values. For example, sigmoid(10) = 0.99995 is practically 1. So instead of using the function every time, I created a static array of pre-calculated sigmoid values, and a method that sets it equal to 0 or 1 if the values are outside of the range of (-10,10). This sped up the program _significantly_, such that you can do about the same with bigger brains, and even using an array of 100 precalculated values is still good enough to where you could switch the method I made to the actual sigmoid function, and the creatures would still behave the same. 

2. **Improved Brain**:
I cleaned up the code for the brain so that you can add more eyes, memory neurons, constant neurons (not that you'd need more constant neurons) and other neurons so that the brain's labels update accordingly, and so that the creature methods, such as eat and see, and properties (mouth, eyes, etc.) will still correspond to the correct neurons.
Additionally, I wrote some code that will scale the brain to size. So a bigger brain will be scaled down in size so that you can still see all of the inputs, hidden layers, and outputs. The only problem I have right now is having the axons/lines scale correctly, and if you try pointing at certain neurons at the bottom, you'll get an out of bounds array exception. 

3. **Invisible Creatures Fix**:
This is a bug fix pertaining to issue #52, which I later found it had to do with the weight values becoming very large, which caused some of the neurons to become NaN. The way I fixed this is by setting a limit to how big/small a weight can be. This is based on the idea that there is no difference between 2 - 1 and 36926485392 - 36926485391. 
Another thing that caused this problem is that sometimes a creature would vomit a value of NaN, so the tile's yum value would have NaN added to it, which would cause the tile's value to become NaN. Then, a creature would walk by and eat from that tile, ingesting NaN into it, which would propagate throughout its neurons. So now I have it so that if a creature vomits NaN, just add some arbitrary value to the tile instead. 
After fixing this, I've been able to run very long simulations. I think the longest I've run went on for about 70,000 years (as opposed to ~3,000 years max). 

4. **Controlled Vision**:
Instead of having an arbitrary set of eyes, have the distances and angles of those eyes be determined by the creature itself. One issue I found when implementing this is that the eye distance and angles grew to very large sizes (like, taking up the whole screen at times). So I set an energy penalty for when the eye distance and angles become less than or greater than a certain value (and the larger those values, the more energy lost). 
You can have the creature control just the vision distance, just the angle, or neither, in which case it would behave just as it currently does in the main branch. 

5. **Random Mutations at Birth**:
These mutations are different and only occur after the creature has had 3 (or whatever number you want) number of children. After that, there's a 75% probability of giving birth to a completely random creature, just as if it were randomly generated when maintaining the minimum population. 
The point of this is to allow for other types of creatures that are potentially more fit to take over the current population, to increase diversity, avoiding a monolith population.
To keep track of this, each creature, in addition to generation number, will have a mutation number. They all start with 0, but when new mutated creature is created, that number increases by 1 and all of its offspring will inherit that number if they're non-mutants. In the sort menus, you can also sort by number of mutations, increasing or decreasing.

6. **Choose Where to Generate New Creatures (kind of)**:
If you want to start a population in a specific place, such as an island, you can zoom into the place and the new creatures being created, when maintaining the population minimum, _should_ spawn within that frame of view. However, I've noticed that they spawn a bit outside of that frame of view. 

I think this is pretty much everything. I'll update this list if I remember about anything else. I'll be pushing my changes in a little while, in case anyone wants to check them out.

**Comment by PRabahy** [comment]
I'm not a fan of number 5. If you wanted the same sort of thing, I would be much happier seeing the "maintain population" code being modified to randomly generate a new creature every X intervals. With the change that you describe, it actually puts a very large penalty on having a high birthrate because most random creatures die quickly and it take a lot of energy to reproduce.

**Comment by 10carfernandez** [comment]
Yeah I like your idea. The only problem for me is that it'll be hard to keep track of the different types of creatures. So after so many years, how would we know if the current population is descended from the same founder population, if the current population is one that took over recently, or if the population is descended from a mix of different founder populations? 

I think the very last question could be answered by just looking at the generation numbers. If there's an obvious split in generation numbers, then there are probably different groups. 

I think one option that would probably address both of our concerns is to keep it the way I have it, but to have  the creature lose zero energy when it gives birth to a creature that was mutated. That way, it's easier to keep track of the new mutants and the order they came in.

**Comment by aproximateknowledge** [comment]
the problem you have with emerging monocultures is not a problem of internal programming of the creatures but a reflection of their surroundings. if you have a Petri dish with one kind of terrain and one kind of food then you will end up with one kind of creature. because in that situation only one species will be the fittest. evangents (a similair simulatror found here: [https://github.com/GhengopelALPHA/Evagents] ) has a creature set-up which is very much the same as evolv's, but because it has a proper water mechanic, three food sources, a climate and  a 'hazards' layer in it's Petri dish divergent species become a reality in no time. i really don't think you need a function like the one you describe at 5. in fact i find it to do a disservice, because part of the fun is knowing it all came from one source. one succesfull creature that in the end gave birth to all. 

correct me if i'm wrong, but i'm assuming that the thought process behind evolution simulators like this to get the highest possible complexity as an emergent function of simple rules. this is an endlessly fascinating endeavour and i applaud it. i know nothing of programming, i only know about it in fairly abstract terms, so i can't really hep writing code or anything like that. but i have more then a passing interest in evolution. based on that knowledge i'd say simpler organisms then the ones you're making here have had higher biodiversity then usually get with this simulator at this point. this has little to do with the creatures and everything with their surroundings. you're basically telling a fairly complex lifeform to survive in a situation where it can learn to thrive in just a couple of genereations. change up the situation. challenge them. they can learn so much more right now then they will ever learn from living a Petri dish. give them impassible terrain, a proper water mechanic with an aquatic food source, weather, climate, stuff they can pick up and place, make them lay fragile eggs they must protect, an underground layer which they can dig into to nest. give them more complex problems to solve and they will solve them, and in that process species will diverge. 

ps. for more of my rants surrounding this check out: #77 #64 #94

**Comment by 10carfernandez** [comment]
I don't know for certain what the general thought process behind all/most evolution simulators are. My best guess is to see how well evolution can be simulated, which can very quite a bit. 

As for this specific project, I know that the person who originally created it, carykh, at one point expressed a bit of disappointment/frustration at how dumb/uninteresting the creatures were even after so many generations. 

So my guess is that at least for him, it was intended for it to eventually evolve some very smart creatures that exhibited some level of "intelligent" bug-like behavior by taking advantage of neural networks (as opposed to a diverse group of specialized creatures). 

My goal isn't to have a diverse group of creatures as an end game, but rather to have a group of creatures with very smart brains, which, I believe, can only be accomplished if we have drastically different brains competing against each other. Otherwise, the neurons seem to just reach a certain set of values and stay there. 

I do like your idea of diversifying the environment, because it might also help accomplish what I'm trying to accomplish, and I do have some ideas. This just seems easier for now. 

I saw some of the things you wrote and I do agree with a lot of them. I've actually thought about having the creatures do some kind of pheromone excretion before. It's tempting to try all of these implementations, but there are still many bugs to fix before doing something big. That's why I listed these things out so the contributors can decide what's most important right now. 

I think once we start getting the code more organized, all of the things you've listed will be more possible. Otherwise, if we did it now, it'd be very messy and confusing code.
